libImaging/Storage.c

```diff
@@ -265,50 +264,178 @@ ImagingDelete(Imaging im)
 /* ------------------ */
 /* Allocate image as an array of line buffers. */
 
+struct ImagingMemoryArena ImagingDefaultArena = {
+    1,                   // alignment
+    1*1024*1024,         // block_size
+    0,                   // blocks_max
+    0,                   // blocks_cached
+    NULL,                // blocks
+    0, 0, 0, 0, 0        // Stats
+};
+
+int
+ImagingMemorySetBlocksMax(ImagingMemoryArena arena, int blocks_max)
+{
+    void *p;
+    /* Free already cached blocks */
+    ImagingMemoryClearCache(arena, blocks_max);
+
+    if (blocks_max == 0 && arena->blocks != NULL) {
+        free(arena->blocks);
+        arena->blocks = NULL;
+    } else if (arena->blocks != NULL) {
+        p = realloc(arena->blocks, sizeof(*arena->blocks) * blocks_max);
+        if ( ! p) {
+            // Leave previous blocks_max value
+            return 0;
+        }
+        arena->blocks = p;
+    } else {
+        arena->blocks = calloc(sizeof(*arena->blocks), blocks_max);
+        if ( ! arena->blocks) {
+            return 0;
+        }
+    }
+    arena->blocks_max = blocks_max;
+
+    return 1;
+}
+
+void
+ImagingMemoryClearCache(ImagingMemoryArena arena, int new_size)
+{
+    while (arena->blocks_cached > new_size) {
+        arena->blocks_cached -= 1;
+        free(arena->blocks[arena->blocks_cached].ptr);
+        arena->stats_freed_blocks += 1;
+    }
+}
+
+ImagingMemoryBlock
+memory_get_block(ImagingMemoryArena arena, int requested_size, int dirty)
+{
+    ImagingMemoryBlock block = {NULL, 0};
+
+    if (arena->blocks_cached > 0) {
+        // Get block from cache
+        arena->blocks_cached -= 1;
+        block = arena->blocks[arena->blocks_cached];
+        // Reallocate if needed
+        if (block.size != requested_size){
+            block.ptr = realloc(block.ptr, requested_size);
```

Comment:
Is this likely to be realloc'ing on every block, unless you're processing a bunch of images of identical bit width?

Reply:
I believe there are three possible results of reallocating:

on the reduction: most implementations will just decrease several numbers in metadata without actually reallocation
on the expansion when space is enough:  as above
on the expansion when here is no space: actual reallocation on the new address

In theory, the more time program is running the less actual relocations should occur, because all blocks will have enough space ahead after several reallocations.

According to my tests, when blocks_max is enough, only small amount of the reusings lead to actual reallocations (when the new ptr != old ptr).

The situation should improved with more intelligent block selection, not just the topmost one.